Add new Admin UI using data store instead of post APIs

thenbrent opened the pull request:
This is a redo of PR #88 for Issue #85. I'm opening a new PR as #88 has become unwieldy in its history. This PR has merged `1-6-dev` (I couldn't rebase without losing subtree data for the pp list table), and adds additional commits to add sortable columns, display Claim ID and a few other patches to bring the admin UI closer to completeness. The only feature remaining for it to have feature parity with the old admin UI and close #85 is a _Search_ box.

This PR also negates #110, which attempted to solve the problem of accessing an actions data by moving that data into the action itself. The changes required to achieve that are excessive, so I've opted to return to the approach of using the data store for accessing that data, as it is much cleaner for now. It also avoids issues like whether data on the action should be changed, e.g. if it's a _pending_ action's status should be changed to _completed_ after `execute()` is called - something which makes sense, but which is an entirely new paradigm for Action Scheduler, and which may have unforeseen implications.

This PR is ready for review, but I'd encourage you to avoid reviewing the commit history, as it's still unwieldy.

thenbrent commented:
@JPry I've realised there's still a bit more work to do here, so there will be some rebasing and force-pushing going on today. Feel free to ignore for another day if you want to avoid seeing commits come-and-go. 🙂

thenbrent commented:
@JPry this is now ready for review. 🙂

JPry commented on classes/ActionScheduler_LogEntry.php:
```diff
@@ -4,12 +4,31 @@
-	protected $message = '';
+	protected $action_id =  '';
+	protected $message =  '';
+	protected $date;
```
These should have docblocks including `@var` tags.

JPry commented on classes/ActionScheduler_LogEntry.php:
```diff
@@ -4,12 +4,31 @@
+	/**
+	 * Constructor
+	 *
+	 * @param mixed  $action_id	Action ID
```
There should be a space instead of a tab between `$action_id` and `Action ID`. Also, is there a reason the type is "mixed" instead of "int"?

thenbrent replied:
> is there a reason the type is "mixed" instead of "int"?

I assume it's because there is no guarantee an ID will always be an `int`. It is an `int` with built-in datastores, but it could just as easily be a string if a hash or SHA was used to refer to an action by other data stores.

JPry replied:
Ah-hah, makes perfect sense. Thanks 🙂

JPry commented on classes/ActionScheduler_LogEntry.php:
```diff
@@ -4,12 +4,31 @@
+		return $this->date;
 	}
 
 	public function get_action_id() {
```
Let's go ahead and add docblocks to the other methods here.

thenbrent replied:
I've opened https://github.com/Prospress/action-scheduler/issues/125 to address this.

Code review isn't the time to raise or address issues in code that is not being added/modified by the PR.

JPry replied:
I'm completely fine with doing docblocks separately. I've mainly mentioned them here just to make sure everything is covered. Personally, I find it easier to include simple things like docblocks when I'm already working in a file, but I can understand that's not the same for everyone.

JPry commented on classes/ActionScheduler_QueueRunner.php:
```diff
@@ -44,7 +44,7 @@ public function run() {
 		$this->run_cleanup();
 		$count = 0;
-		if ( $this->store->get_claim_count() < apply_filters( 'action_scheduler_queue_runner_concurrent_batches', 5 ) ) {
+		if ( $this->store->get_claim_count() < $this->get_concurrent_batch_count() ) {
```
🎉

JPry commented on classes/ActionScheduler_QueueRunner.php:
```diff
@@ -53,6 +53,10 @@ public function run() {
 		return $count;
 	}
 
+	public function get_concurrent_batch_count() {
```
Would it make sense to move this to the `ActionScheduler_Abstract_QueueRunner` class? I think other queue runners could benefit from it.

thenbrent replied:
@JPry yep for sure. I'll also rename it to `get_allowed_concurrent_batches()` to avoid any ambiguity - the current `get_concurrent_batch_count()` could be confused with returning the number of batches in-progress rather than the maximum number allowed to run concurrently.

JPry commented on classes/ActionScheduler_Store.php:
```diff
@@ -120,12 +127,37 @@
+	/**
+	 * @return mixed
+	 */
+	abstract public function get_claim_id( $action_id );
```
This needs the parameter included in the docblock. Also, should this be a defined type, such as an int?

JPry commented on classes/ActionScheduler_Store.php:
```diff
@@ -120,12 +127,37 @@
 	 */
 	abstract public function find_actions_by_claim_id( $claim_id );
 
+	protected function validate_sql_comparator( $comp ) {
```
This method should include a docblock

JPry commented on classes/ActionScheduler_Store.php:
```diff
@@ -120,12 +127,37 @@
 	abstract public function find_actions_by_claim_id( $claim_id );
 
+	protected function validate_sql_comparator( $comp ) {
+		if ( in_array($comp, array('!=', '>', '>=', '<', '<=', '=')) ) {
```
This needs some formatting adjustments to meet with WP standards.

I think it would be preferable to use a key-based array, but it also looks like this will typically only be used a couple times during a given request, so the performance benefit would be minimal. Meaning, it's also OK to leave as-is 🙂

JPry commented on classes/ActionScheduler_wpPostStore.php:
```diff
@@ -93,7 +93,7 @@ protected function save_action_group( $post_id, $group ) {
 	public function fetch_action( $action_id ) {
 		$post = $this->get_post( $action_id );
-		if ( empty($post) || $post->post_type != self::POST_TYPE || $post->post_status == 'trash' ) {
+		if ( empty($post) || $post->post_type != self::POST_TYPE ) {
```
Since this line is already being modified, the `empty($post)` statement should be formatted for WP standards

thenbrent replied:
@JPry my personal preference is generally to maintain internal consistency rather than to apply progressive enhancement. That way, when you're working with a code base, you're not being presented with multiple competing standards. For me, that avoids interrupting my thinking on the more important parts of the code.

I'm OK with doing a widespread standards update on AS though (unlike WP core). It's not really something I feel the need to do, but if you want to do it, I'll happily review/merge the PR.

JPry replied:
> For me, that avoids interrupting my thinking on the more important parts of the code.

As far as I'm concerned, this is a completely valid reason to avoid making these changes now and leave it as a separate PR like you mentioned. For me, making these kinds of changes as I go is trivial and doesn't interrupt my flow, since I can highlight any relevant lines and use a single keyboard shortcut to reformat according to WP standards. But as I said in another comment, I understand that's not the same for everyone.

JPry commented on classes/ActionScheduler_wpPostStore.php:
```diff
@@ -119,12 +119,64 @@ protected function make_action_from_post( $post ) {
+
+		switch ( $post_status ) {
+			case 'any' :
+				$post_status = 'all';
```
Copy/pasta? Should this be `$action_status`?

JPry commented on classes/ActionScheduler_wpPostStore.php:
```diff
@@ -119,12 +119,64 @@ protected function make_action_from_post( $post ) {
+	/**
+	 * @param string $post_status
+	 *
+	 * @return string
```
This should include an `@throws` tag indicating what causes the `InvalidArgumentException` to be thrown.

JPry commented on classes/ActionScheduler_wpPostStore.php:
```diff
@@ -119,12 +119,64 @@ protected function make_action_from_post( $post ) {
+	/**
+	 * @param string $action_status
+	 *
+	 * @return string
```
This should include an `@throws` tag indicating what causes the `InvalidArgumentException` to be thrown.

JPry commented on classes/ActionScheduler_wpPostStore.php:
```diff
@@ -207,7 +261,8 @@ public function query_actions( $query = array() ) {
 		/** @var wpdb $wpdb */
 		global $wpdb;
-		$sql = "SELECT p.ID FROM {$wpdb->posts} p";
+		$sql  = ( 'count' === $select_or_count ) ? 'SELECT count(p.ID)' : 'SELECT p.ID ';
```
I think it would be a good idea to validate the value of the `$select_or_count` variable, since there are only 2 valid values.

JPry commented on classes/ActionScheduler_wpPostStore.php:
```diff
@@ -265,44 +313,84 @@ public function query_actions( $query = array() ) {
+
+		$sql = $this->get_query_actions_sql( $query, $query_type );
+
+		return ( 'count' === $query_type ) ? $wpdb->get_var( $sql ) : $wpdb->get_col( $sql );
```
`$wpdb->get_var()` returns a string, not an array. This should either be converted to an array, or the docblock should be updated to indicate that it can return either type.

JPry commented on classes/ActionScheduler_wpPostStore.php:
```diff
@@ -466,30 +554,24 @@ public function mark_failure( $action_id ) {
 		}
 	}
 
+	public function get_claim_id( $action_id ) {
```
This should include a docblock

JPry commented on classes/ActionScheduler_wpPostStore.php:
```diff
@@ -466,30 +554,24 @@ public function mark_failure( $action_id ) {
+	public function get_claim_id( $action_id ) {
+		return $this->get_post_column( $action_id, 'post_password' );
+	}
+
```
Let's add a docblock here too, including an `@throws` tag.

JPry commented on classes/ActionScheduler_wpPostStore.php:
```diff
@@ -466,30 +554,24 @@ public function mark_failure( $action_id ) {
+		return $this->get_action_status_by_post_status( $status );
+	}
+
+	private function get_post_column( $action_id, $column_name ) {
```
This should include a docblock.

While this is private and won't be used outside of this class, it may be worth considering whether we want to do any kind of validation on the column name.

thenbrent replied:
I'm going to remove this method instead and use `get_post_field()` in its place. That gets us a bit of extra WP magic sauce, even if it will be less efficient than a `$wpdb` query.

thenbrent commented on classes/ActionScheduler_AdminView.php:
```diff
@@ -28,42 +26,63 @@ public static function instance() {
+	 * Registers action-scheduler under Tools.
+	 *
+	 * This method is called if woocommerce is not activated and we can't register our page
+	 * to WooCommerce's System status page.
```
A good example of why docblocks should explain _why_ a method exists and/or does what it does, not _what_ it does. 👍

thenbrent commented on classes/ActionScheduler_ActionFactory.php:
```diff
@@ -4,6 +4,37 @@
+
+		$action = new $action_class( $hook, $args, $schedule, $group );
+
+		return apply_filters( 'action_scheduler_stored_action_instance', $action, $hook, $args, $schedule, $group );
```
I've added one with SHA: 2a5dca27cc86e7892dd988e13732e51d6f448699. 

It feels quite redundant amidst the code though, because variable names, type declarations on the method and general understanding of what `apply_filters()` does, it feels like the docblock is just repeating the code.

Do you want to see these mostly for use with generated documentation? (i.e. docs outside the code)

JPry replied:
> Do you see these as valuable mainly for use with generated documentation? (i.e. docs outside the code) Or is it for something else?

It would certainly be useful for generated documentation, and I do hope we can make use of that. But I also think it's valuable to be consistent 🤷‍♂️

thenbrent commented on classes/ActionScheduler_ListTable.php:
```diff
@@ -0,0 +1,517 @@
+	 * Prints the logs entries inline. We do so to avoid loading Javascript and other hacks to show it in a modal.
+	 *
+	 * @param array $row Action array.
+	 */
```
@JPry it doesn't return anything

thenbrent commented on classes/ActionScheduler_ListTable.php:
```diff
@@ -0,0 +1,517 @@
+	 *
+	 * @param array $row Action array.
+	 */
+	public function column_log_entries( array $row ) {
```
@JPry why? Are you confusing it with `print_log_entry()`?

JPry replied:
Regarding this and [my other comment](https://github.com/Prospress/action-scheduler/pull/113#discussion_r174603300), this actually goes back to the `WP_List_Table` class itself. Inside `WP_List_Table::single_row_columns()`, these methods are called thanks to this `elseif` block (`*` added to indicate the appropriate line):

```php
} elseif ( method_exists( $this, 'column_' . $column_name ) ) {
	echo "<td $attributes>";
*	echo call_user_func( array( $this, 'column_' . $column_name ), $item );
	echo $this->handle_row_actions( $item, $column_name, $primary );
	echo "</td>";
}
```

The returned values of the method for each row are immediately echoed by the calling function. Also, the `ActionScheduler_ListTable::column_args()` method returns a string rather than calling `echo` directly. This is why I was saying that the generated strings should be returned instead of echoed.

As for the black magic of `PP_List_Table`, here's the order of calls that ultimately lead to each column being called:

* `PP_List_Table::display()`
* `WP_List_Table::display()`
* `WP_List_Table::display_rows_or_placeholder()`
* `WP_List_Table::display_rows()`
* `WP_List_Table::single_row()` (called within `foreach` loop for all rows)
* `WP_List_Table::single_row_columns()`
* `echo call_user_func( array( $this, 'column_' . $column_name ), $item );` (called within a `foreach` loop for with all columns)

thenbrent replied:
@JPry thanks for the explanation. That means some other `column_` methods need to be updated too, both in here and `PP_List_Table`. I'll get that fixed up now.

thenbrent commented on classes/ActionScheduler_ListTable.php:
```diff
@@ -0,0 +1,517 @@
+	 * @param array $row Action array.
+	 */
+	public function column_log_entries( array $row ) {
+		echo '<ol>';
```
@JPry log entries will never be empty because an entry is set when an action is created. So if an action exists, it will have at least one log entry.

JPry replied:
> So if an action exists, it will have at least one log entry.

Good to know, thanks for the clarification

thenbrent commented on classes/ActionScheduler_ListTable.php:
```diff
@@ -0,0 +1,517 @@
+		);
+
+		parent::__construct( array(
+			'singular' => __( 'action-scheduler', 'action-scheduler' ),
```
@cesar-prospress can you explain this one?

JPry replied:
My comment about not translating them came from looking at the constructor for `WP_List_Table`. While I don't *think* that translating these will break anything, it looks like they're meant to be used as slugs, so translation doesn't seem appropriate.

thenbrent replied:
@JPry thanks for explaining. That makes sense. It's not entirely clear from the docblock on `WP_List_Table::__construct()` if they should be keys or translated strings, it does say they'll be used for things like class names, and given `sanitize_key()` is applied to them in `WP_List_Table::__construct()`, I think you're right so I've changed them. 👍

thenbrent commented:
@JPry thanks for the review, I've addressed all the comments (mostly that means adding patches, occasionally it means explaining why the proposed change isn't suitable or asking for more information/clarification about why the proposed change is needed).

I'll probably do more work on this next week to try to clean up the git history a bit more (i.e. rebase all those whitespace/docblock changes into originating commits).

JPry commented:
@thenbrent I've responded to all of the comments that needed a response, and hopefully clarified some of my more confusing remarks. Let me know when you're ready for a re-review of the code.

thenbrent commented:
@JPry thanks for the updates. I've addressed those final issues now. I've also changed the base of this to `master` now that `1-6-dev` has been merged into `master` and I've merged `master` into it to make sure it will merge cleanly. The commit history is still pretty gross, but I don't think I'll be able to spend the time cleaning it up.

JPry commented:
@thenbrent Looks like this PR has conflicts in the `ActionScheduler_wpPostStore` class.

thenbrent commented:
I've just added some more commits to this monstrosity. See 591f664...6925e1d. The most notable items are:

* fixes to display of the next schedule date, and row actions on pending actions
* addition of a search box
* removal of the `'all'` status concept

I've also extended the `PP_List_Table` class and cleaned it up a little so that it takes care of a bit more common work, like displaying status filters.

While the code (and docblocks etc.) here could still use a lot of cleaning up, it's ready for review and with the addition of the _Search_ box, it now checks off all items on #85.

thenbrent commented on vendor/pp-list-table/src/class-pp-list-table.php:
```diff
@@ -0,0 +1,659 @@
+	 * Display the table heading and search query, if any
+	 */
+	protected function display_header() {
+		echo '<h1 class="wp-heading-inline">' . esc_attr( $this->table_header ) . '</h1>';
```
@JPry yeah that's outside the scope of this PR, you can submit an issue at https://github.com/Prospress/pp-list-table

JPry replied:
@thenbrent Could you please give me access to that repo?

thenbrent commented on classes/ActionScheduler_ListTable.php:
```diff
@@ -0,0 +1,518 @@
+	protected function maybe_render_actions( $row, $column_name ) {
+		if ( 'pending' === strtolower( $row['status'] ) ) {
+			return parent::maybe_render_actions( $row, $column_name );
+		}
```
@JPry `parent::maybe_render_actions()`, which is `PP_List_Table::maybe_render_actions()` calls `return;` if `empty( $this->row_actions[ $column_name ] )`, which means it returns `null`, which is the same as not calling `return` I believe. Happy for that to be improved if you want, it's pretty sloppy/unnecessarily inconsistent, but I don't see a need for that to block this PR being merged.

JPry replied:
>  Happy for that to be improved if you want, it's pretty sloppy/unnecessarily inconsistent, but I don't see a need for that to block this PR being merged.

It's no longer blocking, as I included it in the updates I made to the PR while you were out 🙂

thenbrent commented on classes/ActionScheduler_ListTable.php:
```diff
@@ -0,0 +1,518 @@
+	 */
+	protected function get_recurrence( $action ) {
+		if ( $action->get_schedule()->is_recurring() ) {
+			return sprintf( __( 'Every %s', 'action-scheduler' ), self::human_interval( $action->get_schedule()->interval_in_seconds() ) );
```
@JPry that is beyond the scope fo this PR. It's a legitimate issue though, so feel free to open a new Issue for it.

thenbrent replied:
@JPry I've added https://github.com/Prospress/action-scheduler/issues/133 to address the broader issue with that.

To address the specific issue in this PR, I've reverted https://github.com/Prospress/action-scheduler/pull/113/commits/73fcd9e0bfb936602c513dadce637fb8b77ac7fa.

Not ideal using `method_exists()`, but until #133 is implemented, it's the only reliable approach (I don't want to add a fix for #133 to this already bloated PR).

thenbrent commented:
@JPry thanks for finding/fixing those issues! Commits https://github.com/Prospress/action-scheduler/compare/6925e1df...3a9ff9ec6a all look good to me.

> Just calling out that this is still here instead of the `get_post_field()` function mentioned above.

Not sure why, but GitHub won't let me reply inline to that comment.

I had to remove the use of `get_post_field()`. For an explanation of why, see comments on https://github.com/Prospress/action-scheduler/pull/113/commits/f3c76e7980ac8f8bc692b484eae6980dc717b4ed
